### PR TITLE
Suppress exceptions during expression optimization

### PR DESCRIPTION
**Pull request by kasiafi**

### CODE REVIEW

File: core/trino-main/src/test/java/io/trino/type/TestArrayOperators.java [hunk 1]
```diff
@@ -312,17 +312,17 @@ public void testJsonToArray()
...
-        assertInvalidCast("CAST(JSON '[1, 2, 3]' AS ARRAY<ARRAY<BIGINT>>)", "Cannot cast to array(array(bigint)). Expected a json array, but got 1\n[1,2,3]");
-        assertInvalidCast("CAST(JSON '[1, {}]' AS ARRAY<BIGINT>)", "Cannot cast to array(bigint). Unexpected token when cast to bigint: {\n[1,{}]");
-        assertInvalidCast("CAST(JSON '[[1], {}]' AS ARRAY<ARRAY<BIGINT>>)", "Cannot cast to array(array(bigint)). Expected a json array, but got {\n[[1],{}]");
+        assertInvalidCast("CAST(JSON '{\"a\": 1}' AS ARRAY<BIGINT>)", "Cannot cast to array(bigint). Expected a json array, but got {\n{\"a\": 1}");
+        assertInvalidCast("CAST(JSON '[1, 2, 3]' AS ARRAY<ARRAY<BIGINT>>)", "Cannot cast to array(array(bigint)). Expected a json array, but got 1\n[1, 2, 3]");
+        assertInvalidCast("CAST(JSON '[1, {}]' AS ARRAY<BIGINT>)", "Cannot cast to array(bigint). Unexpected token when cast to bigint: {\n[1, {}]");
+        assertInvalidCast("CAST(JSON '[[1], {}]' AS ARRAY<ARRAY<BIGINT>>)", "Cannot cast to array(array(bigint)). Expected a json array, but got {\n[[1], {}]");
```
Comment (martint) [inline core/trino-main/src/test/java/io/trino/type/TestArrayOperators.java:318]: Unrelated formatting change

File: core/trino-main/src/test/java/io/trino/type/TestArrayOperators.java [hunk 2]
```diff
@@ -312,17 +312,17 @@ public void testJsonToArray()
...
 
         assertInvalidCast("CAST(unchecked_to_json('1, 2, 3') AS ARRAY<BIGINT>)", "Cannot cast to array(bigint).\n1, 2, 3");
         assertInvalidCast("CAST(unchecked_to_json('[1] 2') AS ARRAY<BIGINT>)", "Cannot cast to array(bigint). Unexpected trailing token: 2\n[1] 2");
         assertInvalidCast("CAST(unchecked_to_json('[1, 2, 3') AS ARRAY<BIGINT>)", "Cannot cast to array(bigint).\n[1, 2, 3");
 
-        assertInvalidCast("CAST(JSON '[\"a\", \"b\"]' AS ARRAY<BIGINT>)", "Cannot cast to array(bigint). Cannot cast 'a' to BIGINT\n[\"a\",\"b\"]");
-        assertInvalidCast("CAST(JSON '[1234567890123.456]' AS ARRAY<INTEGER>)", "Cannot cast to array(integer). Out of range for integer: 1.234567890123456E12\n[1.234567890123456E12]");
+        assertInvalidCast("CAST(JSON '[\"a\", \"b\"]' AS ARRAY<BIGINT>)", "Cannot cast to array(bigint). Cannot cast 'a' to BIGINT\n[\"a\", \"b\"]");
+        assertInvalidCast("CAST(JSON '[1234567890123.456]' AS ARRAY<INTEGER>)", "Cannot cast to array(integer). Out of range for integer: 1.234567890123456E12\n[1234567890123.456]");
```
Comment (martint) [inline core/trino-main/src/test/java/io/trino/type/TestArrayOperators.java:325]: Why did this change?

File: core/trino-main/src/test/java/io/trino/type/TestMapOperators.java [hunk 3]
```diff
@@ -445,20 +445,20 @@ public void testJsonToMap()
...
-        assertInvalidCast("CAST(JSON '{\"1\": {\"a\": 1}, \"2\": []}' AS MAP<VARCHAR, MAP<VARCHAR, BIGINT>>)", "Cannot cast to map(varchar, map(varchar, bigint)). Expected a json object, but got [\n{\"1\":{\"a\":1},\"2\":[]}");
+        assertInvalidCast("CAST(JSON '[1, 2]' AS MAP<BIGINT, BIGINT>)", "Cannot cast to map(bigint, bigint). Expected a json object, but got [\n[1, 2]");
+        assertInvalidCast("CAST(JSON '{\"a\": 1, \"b\": 2}' AS MAP<VARCHAR, MAP<VARCHAR, BIGINT>>)", "Cannot cast to map(varchar, map(varchar, bigint)). Expected a json object, but got 1\n{\"a\": 1, \"b\": 2}");
+        assertInvalidCast("CAST(JSON '{\"a\": 1, \"b\": []}' AS MAP<VARCHAR, BIGINT>)", "Cannot cast to map(varchar, bigint). Unexpected token when cast to bigint: [\n{\"a\": 1, \"b\": []}");
+        assertInvalidCast("CAST(JSON '{\"1\": {\"a\": 1}, \"2\": []}' AS MAP<VARCHAR, MAP<VARCHAR, BIGINT>>)", "Cannot cast to map(varchar, map(varchar, bigint)). Expected a json object, but got [\n{\"1\": {\"a\": 1}, \"2\": []}");
```
Comment (martint) [inline core/trino-main/src/test/java/io/trino/type/TestMapOperators.java:451]: Unrelated formatting changes

File: testing/trino-tests/src/test/java/io/trino/tests/AbstractTestEngineOnlyQueries.java [hunk 4]
```diff
@@ -5810,10 +5810,10 @@ public void testLargePivot()
     @Test
     public void testPivotExceedingMaximumArraySize()
     {
-        assertQueryFails(pivotQuery(255), "Too many arguments for array constructor");
+        assertQueryFails(pivotQuery(255, false), "Too many arguments for array constructor");
     }
 
-    private static String pivotQuery(int columnsCount)
+    private static String pivotQuery(int columnsCount, boolean interpretable)
```
Comment (martint) [inline testing/trino-tests/src/test/java/io/trino/tests/AbstractTestEngineOnlyQueries.java:5816]: Not sure I understand why this is needed

File: core/trino-main/src/test/java/io/trino/operator/scalar/TestConditions.java [hunk 5]
```diff
@@ -157,12 +156,11 @@ public void testIn()
...
-    }
 
-    @Test
-    public void testInDoesNotShortCircuit()
-    {
-        assertInvalidFunction("3 in (2, 4, 3, 5 / 0)", DIVISION_BY_ZERO);
+        // Short circuit: this expression is optimized by SimplifyExpressions to 'true', so it does not fail due to division by zero.
+        // However, the test fails, because one of it parts is "interpret", which tries to evaluate the unoptimized expression
+        // and throws DIVISION_BY_ZERO.
+        /*assertFunction("3 in (2, 4, 3, 5 / 0)", BOOLEAN, true);*/
```
[the remark this answers is not in the document]
Comment (kasiafi) [reply core/trino-main/src/test/java/io/trino/operator/scalar/TestConditions.java:163]: `SimplifyExpressions` optimizes it to `true`.
This is done by `ExpressionInterpreter.optimize()`. It indeed evaluates all expressions in the IN-list, but error is suppressed, and eventually the match is found.
Is it incorrect that it succeeds?

File: core/trino-main/src/test/java/io/trino/type/TestMapOperators.java (hunk 3 above)
Comment (kasiafi) [reply core/trino-main/src/test/java/io/trino/type/TestMapOperators.java:451]: They are related. 
The previous version of the message came from the "wrapping" failure function, and it went through json coding / decoding.
The new version of the message is the "raw" version. It comes directly from the failure.

The change in numeric format you pointed in another comment has the same reason.

File: core/trino-main/src/test/java/io/trino/sql/TestExpressionInterpreter.java [hunk 6]
```diff
@@ -1569,10 +1570,7 @@ private static Object evaluate(Expression expression)
         public Expression rewriteFunctionCall(FunctionCall node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
         {
             if (extractFunctionName(node.getName()).equals("fail")) {
-                return new FunctionCallBuilder(METADATA)
-                        .setName(QualifiedName.of("fail"))
-                        .addArgument(VARCHAR, new StringLiteral("fail"))
-                        .build();
```
[the remark this answers is not in the document]
Comment (kasiafi) [reply core/trino-main/src/test/java/io/trino/sql/TestExpressionInterpreter.java:1575]: The issue was that the left-hand-side expression (`actual`) contained the failure function resolved with extra long name, while the right-hand-side expression (`expected`) had the fail function unresolved, so they didn't match.

Anyway, this rewrite is removed in the following commit. I should probably squash them.

File: testing/trino-tests/src/test/java/io/trino/tests/AbstractTestEngineOnlyQueries.java (hunk 4 above)
Comment (kasiafi) [reply testing/trino-tests/src/test/java/io/trino/tests/AbstractTestEngineOnlyQueries.java:5816]: `interpretable` is a way to decide whether the array will be created in `ExpressionInterpreter`.
If all of the array elements can be evaluated, then the Interpreter creates the array and it doesn't fail even if the array is very big. 
I didn't find out though why this test passed before my change.

File: core/trino-main/src/test/java/io/trino/operator/scalar/TestConditions.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (martint) [reply core/trino-main/src/test/java/io/trino/operator/scalar/TestConditions.java:163]: Yes, it's incorrect because the order of evaluation should consider first all the elements in the list and only *then* do the containment check.

File: testing/trino-tests/src/test/java/io/trino/tests/AbstractTestEngineOnlyQueries.java (hunk 4 above)
Comment (kasiafi) [reply testing/trino-tests/src/test/java/io/trino/tests/AbstractTestEngineOnlyQueries.java:5816]: I investigated a bit and here is what I found:
- before, the array constructor failed in `ExpressionInterpreter`
- after I added error suppression, it stopped to fail in `ExpressionInterpreter`, and did not fail at all, because it missed the size check in `ArrayConstructor`. **Probably it is a bug.**
- then I added `0/0` to the arrays and it started failing again with "Too many arguments for array constructor". But it was neither in `ExpressionInterpreter` nor in `ArrayConstructor`. It was **in the `ExpressionAnalyzer`** invoked by `DesugarLikeRewrite`.

I think we need an unconditional size check in the `ArrayConstructor`. We cannot rely on a chance to hit one of the checks we've got here and there.

File: core/trino-main/src/test/java/io/trino/operator/scalar/TestConditions.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (kasiafi) [reply core/trino-main/src/test/java/io/trino/operator/scalar/TestConditions.java:163]: Okay, so we've got a bug. According to what you said, this query should fail.
```
trino:tiny> SELECT * FROM region WHERE 1 IN (1, 2, regionkey / regionkey);
 regionkey |    name     |                                                       comment
-----------+-------------+---------------------------------------------------------------------------------------------------------------------
         0 | AFRICA      | lar deposits. blithely final packages cajole. regular waters are final requests. regular accounts are according to
         1 | AMERICA     | hs use ironic, even requests. s
         2 | ASIA        | ges. thinly even pinto beans ca
         3 | EUROPE      | ly final courts cajole furiously final excuse
         4 | MIDDLE EAST | uickly special accounts cajole carefully blithely close requests. carefully final asymptotes haggle furiousl
(5 rows)
```

[the remark this answers is not in the document]
Comment (kasiafi) [reply core/trino-main/src/test/java/io/trino/operator/scalar/TestConditions.java:163]: I discovered that in case when some in-list expression causes exception, we currently:
- either fail the query in `SimplifyExpression` which invokes `ExpressionInterpreter` (that covered "0/0")
- or the query fails in execution
- or the query does not fail at all.

My change took away the first possibility.
I hoped that if the "failing" in-predicate survives until execution, it will fail there. But that was too optimistic. The case `3 in (2, 4, 3, 5 / 0)` slipped through. 

I looked into `InCodeGenerator`, which is responsible for the execution of in-predicates. It is quite complicated. It is optimized for different lengths of the in-list and for different types. **One of the options, named `DIRECT_SWITCH`, drops the non-constant elements to "default bucket" and evaluates them lazily**.
According to what you said, it is a bug. I think that we should investigate `InCodeGenerator`and make sure that he always evaluates all values.

[the remark this answers is not in the document]
Comment (kasiafi) [reply core/trino-main/src/test/java/io/trino/operator/scalar/TestConditions.java:163]: What should I do regarding this PR? 
It's no good until `InCodeGenerator` is fixed, because it exposes the issue to a great extent.
Or maybe I should add some "sanity checker" to fail in-predicates before execution?

File: core/trino-main/src/test/java/io/trino/sql/TestExpressionInterpreter.java [hunk 7]
```diff
@@ -1498,6 +1498,14 @@ public void testRowSubscript()
     {
         assertOptimizedEquals("ROW(1, 'a', true)[3]", "true");
         assertOptimizedEquals("ROW(1, 'a', ROW(2, 'b', ROW(3, 'c')))[3][3][2]", "'c'");
+
+        assertOptimizedEquals("ROW(1, null)[2]", "null");
+        assertOptimizedEquals("ROW(0 / 0, 1)[1]", "ROW(0 / 0, 1)[1]");
+        assertOptimizedEquals("ROW(0 / 0, 1)[2]", "1");
```
Comment (martint) [inline core/trino-main/src/test/java/io/trino/sql/TestExpressionInterpreter.java:1504]: This is an interesting case. Arguably, this should keep the original expression intact, as the optimized version has different semantics -- the arguments to ROW are expected to be evaluated eagerly, so the expression should fail during evaluation regardless of what optimizations get applied.

Comment (kasiafi) [reply core/trino-main/src/test/java/io/trino/sql/TestExpressionInterpreter.java:1504]: OK. I'll remove the changes related to Row. Are there any other over-optimizations?

File: core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java [hunk 8]
```diff
@@ -696,7 +697,14 @@ protected Object visitArithmeticUnary(ArithmeticUnaryExpression node, Object con
...
             if (value instanceof Expression) {
-                return new ArithmeticUnaryExpression(node.getSign(), toExpression(value, type(node.getValue())));
+                Expression valueExpression = toExpression(value, type(node.getValue()));
+                switch (node.getSign()) {
+                    case PLUS:
+                        return valueExpression;
+                    case MINUS:
+                        return new ArithmeticUnaryExpression(MINUS, valueExpression);
+                }
+                throw new UnsupportedOperationException("Unsupported unary operator: " + node.getSign());
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java:707]:
```
default:
    throw new ...
```

File: core/trino-main/src/test/java/io/trino/sql/TestExpressionInterpreter.java [hunk 9]
```diff
@@ -336,6 +336,17 @@ public void testNegative()
         assertOptimizedEquals("-(bound_decimal_long-12345678901234567890.123)", "CAST(0 AS decimal(20,10))");
     }
 
+    @Test
+    public void testArithmeticUnary()
+    {
+        assertOptimizedEquals("-rand()", "-rand()");
+        assertOptimizedEquals("-(0 / 0)", "-(0 / 0)");
+        assertOptimizedEquals("-(-(-(0 / 0)))", "-(-(-(0 / 0)))");
```
Comment (martint) [inline core/trino-main/src/test/java/io/trino/sql/TestExpressionInterpreter.java:344]: For later, we should add an optimization that removes redundant negative unary expressions. E.g. `-(-(x))` => `x`

Comment (kasiafi) [reply core/trino-main/src/test/java/io/trino/sql/TestExpressionInterpreter.java:344]: Added.

File: core/trino-main/src/test/java/io/trino/sql/query/TestExpressions.java [hunk 10]
```diff
@@ -51,6 +51,12 @@ public void testBooleanExpressionInCase()
         assertThat(assertions.query("VALUES CASE 1 > 2 WHEN true THEN 10 ELSE 20 END")).matches("VALUES 20");
     }
 
+    @Test
+    public void testInShortCircuit()
+    {
+        assertThat(assertions.query("SELECT 3 IN (2, 4, 3, 5 / 0)")).matches("VALUES true");
```
[the remark this answers is not in the document]
Comment (kasiafi) [reply core/trino-main/src/test/java/io/trino/sql/query/TestExpressions.java:57]: I made a change in `ExpressionInterpreter` so that now it doesn't optimize in-predicates by short-circuit.
However, this test still passes, because the generated code for in-predicate does short-circuit. I added comments.

File: core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java [hunk 11]
```diff
@@ -17,10 +17,8 @@
 import com.google.common.collect.ImmutableMap;
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java:17]: Typo in commit message: experssion

File: core/trino-main/src/test/java/io/trino/sql/query/TestExpressions.java [hunk 12]
```diff
@@ -51,6 +51,19 @@ public void testBooleanExpressionInCase()
         assertThat(assertions.query("VALUES CASE 1 > 2 WHEN true THEN 10 ELSE 20 END")).matches("VALUES 20");
     }
 
+    @Test
+    // Because of the failing in-list item 5 / 0, the in-predicate cannot be simplified.
```
Comment (martint) [inline core/trino-main/src/test/java/io/trino/sql/query/TestExpressions.java:55]: Put the comment inside the method
